machida/lib/wallaroo/experimental/__init__.py

```diff
@@ -100,6 +125,346 @@ def write(self, message, event_time=0, key=None):
         self._conn.sendall(payload)
 
 
+Stream = namedtuple('Stream', ['id', 'name', 'point_of_ref', 'is_open'])
+
+
+class AtLeastOnceSourceConnector(threading.Thread, asynchat.async_chat, BaseConnector):
+    # TODO :
+    # 1. implement credits based flow control
+    # 2. revisit parse_connector_args
+    # The way we require host and port args, but then get them from the app topo
+    # and ignore the ones in the parsed params object is confusing and doesn't
+    # make much sense to me. Ping Brian before making changes to this.
+
+    def __init__(self, args=None, required_params=['host', 'port'],
+                 optional_params=['timeout']):
+        # Use BaseConnector to do any argument parsing
+        BaseConnector.__init__(self, args, required_params, optional_params)
+
+        # connection details are given from the base
+        self._port = int(self._port)  # but convert port to int
+        self.credits = 0
+        if self.params.timeout:
+            self._timeout = float(self.params.timeout)
+        else:
+            self._timeout = 0.05
+
+        # Stream details
+        self._streams = {}  # stream_id: Stream
+
+        self.handshake_complete = False
+        self.in_buffer = []
+        self.out_buffer = []
+        self.reading_header = True
+        self.set_terminator(4) # first frame header
+        self.error = None
+
+        # Thread details
+        threading.Thread.__init__(self)
+        self.daemon = True
+
+    ###########################
+    # Incoming communications #
+    ###########################
+
+    def run(self):
+        # only allow starting after handshake is complete
+        if not self.handshake_complete:
+            raise ProtocolError("Cannot start async loop before handshake"
+                                " is complete")
+        asyncore.loop(timeout=self._timeout)
+
+    def collect_incoming_data(self, data):
+        """Buffer the data"""
+        self.in_buffer.append(data)
```

Comment:
It's unbounded and can be anything that's sent to this connection. If wallaroo goes wonky, this could get big.
I'm not sure what the correct bounding size is in here, though. We could also limit the flow by referencing the size of this in `readable()` to return `False` if it's greater than some arbitrary threshold.